carball/analysis/stats/controls/controls.py

```diff
@@ -0,0 +1,37 @@
+from typing import Dict
+
+import pandas as pd
+
+from ....analysis.stats.stats import BaseStat
+from ....generated.api import game_pb2
+from ....generated.api.player_pb2 import Player
+from ....generated.api.stats.player_stats_pb2 import PlayerStats
+from ....json_parser.game import Game
+
+
+class ControlsStat(BaseStat):
+
+    def calculate_player_stat(self, player_stat_map: Dict[str, PlayerStats], game: Game, proto_game: game_pb2.Game,
+                              player_map: Dict[str, Player], data_frame: pd.DataFrame):
+        for player_key, stats in player_map.items():
+            player_name = player_map[player_key].name
+            steer_values = data_frame[player_name]['steer'].tolist()
+            throttle_values = data_frame[player_name]['throttle'].tolist()
+            steering_percentage = self.get_analogue_percentage(steer_values)
+            throttle_percentage = self.get_analogue_percentage(throttle_values)
+            is_keyboard = (steering_percentage == 0) and (throttle_percentage == 0)
+
+            controller_stats = player_stat_map[player_key].controller
+            controller_stats.is_keyboard = is_keyboard
+            controller_stats.analogue_steering_input_percent = throttle_percentage
+            controller_stats.analogue_throttle_input_percent = steering_percentage
+
+
+    def get_analogue_percentage(self, values: list):
+        total_frames = values.__len__()
+        count = 0
+        for value in values:
```

Comment:
Instead of a for loop you can do something like:
```python
(data_frame[player_name]['steer'] == 128).sum()
```
to get the total number of frames where that statement is true (i.e. steer is equal to exactly 128). You can do that for all values (0, 128, 255) and add them together to get the whole percentage. This is much, much faster.

Reply:
Additionally, you don't need to convert to a list for any of this. If you need to convert to a list-like object at some point, use `.values` instead of `tolist`, as numpy arrays are significantly faster and have their own versions of aggregate functions.